src/geometry/manifold/manifold.lean

```diff
@@ -0,0 +1,534 @@
+inherits additional structure (it makes sense to talk about smooth manifolds). There are therefore
+two different ingredients in a manifold:
+* the set of charts, which is data
+* the fact that changes of charts belong to some group (in fact groupoid), which is additional Prop.
+
+We separate these two parts in the definition: the manifold structure is just the set of charts, and
+then the different smoothness requirements (smooth manifold, orientable manifold, contact manifold,
+and so on) are additional properties of these charts. These properties are formalized through the
+notion of structure groupoid, i.e., a set of local homeomorphisms stable under composition and
+inverse, to which the change of coordinates should belong.
+
+## Main definitions
+`structure_groupoid H`   : a subset of local homeomorphisms of `H` stable under composition, inverse
+                           and restriction (ex: local diffeos)
+`pregroupoid H`          : a subset of local homeomorphisms of `H` stable under composition and
+                           restriction, but not inverse (ex: smooth maps)
+`groupoid_of_pregroupoid`: construct a groupoid from a pregroupoid, by requiring that a map and its
+                           inverse both belong to the pregroupoid (ex: construct diffeos from smooth
+                           maps)
+`continuous_groupoid H`  : the groupoid of all local homeomorphisms of `H`
+
+`manifold H M`           : manifold structure on M modelled on H, given by an atlas of local
+                           homeomorphisms from M to H whose sources cover M. This is a type class.
+`has_groupoid M G`       : when `G` is a structure groupoid on `H` and `M` is a manifold modelled on
+                           `H`, require that all coordinate changes belong to `G`. This is a type
+                           class
+`atlas M H`              : when `M` is a manifold modelled on `H`, the atlas of this manifold
+                           structure, i.e., the set of charts
+`diffeomorph G M M'`     : the set of diffeomorphisms between the manifolds M and M' for the
+                           groupoid G
+
+## Implementation notes
+
+The atlas in a manifold is *not* a maximal atlas in general: the notion of maximality depends on the
+groupoid one considers, and changing groupoids changes the maximal atlas. With the current
+formalization, it makes sense first to choose the atlas, and then to ask whether this precise atlas
+defines a smooth manifold, an orientable manifold, and so on.
+
+We use the letter `H` for the model space thinking of the case of manifolds with boundary, where the
+model space is a half space.
+-/
+
+noncomputable theory
+local attribute [instance, priority 0] classical.decidable_inhabited classical.prop_decidable
+
+universes u
+
+variables {H : Type u} {M : Type*} {M' : Type*} {M'' : Type*}
+
+local infixr  ` →ₕ `:100 := local_homeomorph.trans
```

Comment:
This arrow is really weird to me--I always want to read it as some kind of function type. How about something like `≫ₕ`?

Reply:
@sgouezel What do you think of this suggestion?

Reply:
I had missed this comment, thanks for reminding me. Yes, this is much better, change made.